README.md
response-patch
==============

[![Build Status](https://secure.travis-ci.org/fengmk2/response-patch.png)](http://travis-ci.org/fengmk2/response-patch)
monkey patch for `http.ServerResponse`.

## Install

```bash
$ npm install response-patch
```

## Usage

```js
require('response-patch');
var http = require('http');
http.createServer(function (req, res) {
  res.req = req;
  res.send('hello world');
});
```

## License

(The MIT License)

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md

0.1.0 / 2013-05-06 
==================

  * hotfix for XSS in jsonp callback
  * update readme

0.0.3 / 2013-01-08 
==================

  * fixed charset and not change statusCode by jsonp

0.0.2 / 2012-11-23 
==================

  * Merge pull request #1 from jifeng/master
  * number
  * fixed buffer bug
  * add testcase of array and null
  * fix test case bug about 9.0 version of node
  * add version
  * add test cases
  * style and use send function
  * add jsonp function
  * add to travis-ci


test/patch.js
/*!
 * reponse-patch - test/patch.js
 * Copyright(c) 2012 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

"use strict";

/**
 * Module dependencies.
 */

require('../');
var http = require('http');
var request = require('supertest');
var should = require('should');

describe('patch.js', function () {

  var app = http.createServer(function (req, res) {
    res.req = req;

    if (req.url === '/send') {
      return res.send('hello');
    }
    if (req.url === '/send/buffer') {
      return res.send(new Buffer('hello你好'));
    }
    if (req.url === '/send/json') {
      return res.send({foo: 'bar'});
    }
    if (req.url === '/json') {
      return res.json({json: 123});
    }
    if (req.url === '/redirect') {
      return res.redirect('/');
    }
    if (req.url === '/redirect/301') {
      return res.redirect('/301', 301);
    }
    if (req.url === '/redirect/302') {
      return res.redirect('/302', '302');
    }
    if (req.url === '/jsonp') {
      return res.jsonp({key1: 'value1'});
    }
    if (req.url === '/jsonp/xss1') {
      return res.jsonp({key1: 'value1'}, '</script><script>alert(399220605)</script>');
    }
    if (req.url === '/jsonp/xss2') {
      return res.jsonp({key1: 'value1'}, decodeURIComponent('%3C/script%3E%3Cscript%3Ealert(469517365)%3C/script%3E'));
    }
    if (req.url === '/jsonp/cb') {
      return res.jsonp({key2: 'value2'}, 'cb');
    }
    if (req.url === '/jsonp/buffer') {
      return res.jsonp(new Buffer('mockbuffer'));
    }
    if (req.url === '/jsonp/undefined') {
      return res.jsonp();
    }
    if (req.url === '/jsonp/string') {
      return res.jsonp('str');
    }
    if (req.url === '/jsonp/string500') {
      res.statusCode = 500;
      return res.jsonp('str500');
    }
    if (req.url === '/jsonp/number') {
      return res.jsonp(1);
    }
    if (req.url === '/jsonp/array') {
      return res.jsonp([1, 2, 3]);
    }
    if (req.url === '/jsonp/null') {
      return res.jsonp(null);
    }
  });

  it('should send(str)', function (done) {
    request(app)
    .get('/send')
    .expect(200, 'hello')
    .expect('Content-Type', 'text/html', done);
  });

  it('should send(buffer)', function (done) {
    request(app)
    .get('/send/buffer')
    .expect(200)
    .expect('Content-Type', 'application/octet-stream', done);
  });

  it('should send(json)', function (done) {
    request(app)
    .get('/send/json')
    .expect(200, '{"foo":"bar"}')
    .expect('Content-Type', 'application/json', done);
  });

  it('should json(json)', function (done) {
    request(app)
    .get('/json')
    .expect(200, '{"json":123}')
    .expect('Content-Type', 'application/json', done);
  });

  it('should redirect()', function (done) {
    request(app)
    .get('/redirect')
    .expect(302)
    .expect('Location', '/', done);
  });

  it('should redirect(301)', function (done) {
    request(app)
    .get('/redirect/301')
    .expect(301)
    .expect('Location', '/301', done);
  });

  it('should redirect(302)', function (done) {
    request(app)
    .get('/redirect/302')
    .expect(302)
    .expect('Location', '/302', done);
  });

  it('should jsonp {"key1": "value1"}', function (done) {
    request(app)
    .get('/jsonp')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback({"key1":"value1"})', done);
  });

  it('should jsonp {"key2": "value2"}', function (done) {
    request(app)
    .get('/jsonp/cb')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('cb({"key2":"value2"})', done);
  });

  it('should jsonp [1 ,2, 3]', function (done) {
    request(app)
    .get('/jsonp/array')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback([1,2,3])', done);
  });
 
  it('should error response data when data is buffer', function (done) {
    request(app)
    .get('/jsonp/buffer')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback("mockbuffer")', done);
  });

  it('should when data and callback both is undefined', function (done) {
    request(app)
    .get('/jsonp/undefined')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback(undefined)', done);
  });

  it('should jsonp response data when data is string', function (done) {
    request(app)
    .get('/jsonp/string')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback("str")', done);
  });

  it('should jsonp response data type is string and status 500', function (done) {
    request(app)
    .get('/jsonp/string500')
    .expect(500)
    .expect('Content-Type', 'application/javascript')
    .expect('callback("str500")', done);
  });

  it('should jsonp response data when data is number', function (done) {
    request(app)
    .get('/jsonp/number')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback(1)', done);
  });
  
  it('should jsonp response data when data is null', function (done) {
    request(app)
    .get('/jsonp/null')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('callback(null)', done);
  });

  it('should jsonp callback protect XSS1', function (done) {
    request(app)
    .get('/jsonp/xss1')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('&lt;/script&gt;&lt;script&gt;alert(399220605)&lt;/script&gt;({"key1":"value1"})', done);
  });

  it('should jsonp callback protect XSS2', function (done) {
    request(app)
    .get('/jsonp/xss2')
    .expect(200)
    .expect('Content-Type', 'application/javascript')
    .expect('&lt;/script&gt;&lt;script&gt;alert(469517365)&lt;/script&gt;({"key1":"value1"})', done);
  });
});


lib/patch.js
/*!
 * response-patch - lib/patch.js
 * Copyright(c) 2012 Taobao.com
 * Author: suqian.yf <suqian.yf@taobao.com>
 */

"use strict";

/**
 * Module dependencies.
 */

var utility = require('utility');
var http = require('http');
var send = require('response-send');
var res = http.ServerResponse.prototype;

// #nodejsWTF? not work in node < 0.8.0
// https://github.com/visionmedia/node-response-send/issues/1

// http.ServerResponse.prototype.__defineGetter__('req', function () {
//   return this.__req || this.socket.parser.incoming;
// });

/**
 * Send a response.
 *
 * Examples:
 *
 *     res.send(new Buffer('wahoo'));
 *     res.send({ some: 'json' });
 *     res.send('<p>some html</p>');
 *     res.send(404, 'Sorry, cant find that');
 *     res.send(404);
 *
 * @param {Mixed} body or status
 * @param {Mixed} body
 * @return {ServerResponse}
 * @api public
 */
res.send = send;

res.json = send.json();

/**
 * Send redirect response.
 * 
 * @param  {Response} res, http.Response instance
 * @param  {String} url, redirect URL
 * @param  {Number|String} status, response status code, default is `302`
 * @api public
 */
res.redirect = function (url, status) {
  if (typeof status === 'string') {
    status = parseInt(status, 10);
  }
  this.statusCode = status === 301 ? 301 : 302;
  this.setHeader('Location', url);
  this.end();
};

/**
 * Jsonp response.
 * 
 * @param {Object} data, object is only supported
 * @param {String} [callback], default is `'callback'`.
 * Examples:
 *
 * ```js
 * res.jsonp({"key": "value"});
 * res.jsonp({"key": "value"}, 'cb');
 * ```
 * 
 * @api public
 */
res.jsonp = function (data, callback) {
  callback = callback || 'callback';
  if (Buffer.isBuffer(data)) {
    data = data.toString();
  }
  var body = JSON.stringify(data);
  body = utility.escape(callback) + '(' + body + ')';
  this.charset = this.charset || 'utf-8';
  this.setHeader('Content-Type', 'application/javascript');
  return this.send(body); 
};
